### PR TITLE
Glooctl check improvements

### PR DESCRIPTION
**Pull request by arianaw66**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1627

### CODE REVIEW

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 1]
```diff
@@ -360,6 +368,94 @@ func checkProxies(namespaces []string) (bool, error) {
...
+	}
+
+	desiredMetricsSegments := []string{"update_attempt", "update_rejected", "update_failure"}
+	statsMap := parseMetrics(stats, desiredMetricsSegments, promStatsPath)
+	newStatsMap := parseMetrics(newStats, desiredMetricsSegments, promStatsPath)
+
+	if reflect.DeepEqual(newStatsMap, statsMap) {
+		// TODO Is worth the time for reflect?
+		fmt.Println("same though")
+		fmt.Printf("OK\n")
```
Comment (ashleywang1) [inline projects/gloo/cli/pkg/cmd/check/root.go]: It's glooctl check, so I think it's worth it.
The "same though" output should probably be removed.

Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: Haha yeah. Was just leaving it there while there's a TODO. And thanks! Meant to make a comment that I'd like feedback on the TODO

File: pkg/cliutil/uri.go [hunk 2]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
 
 	return strings.TrimSuffix(hostname.String(), "\n"), err
 }
+
+// Call kubectl port-forward. Callers are expected to clean up the returned portFwd *exec.cmd after the port-forward is no longer needed.
+func PortForward(namespace string, resource string, localPort string, kubePort string, verbose bool) (error, *exec.Cmd) {
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: by convention `error` is the second return from a function

File: changelog/v1.3.13/improve-glooctl-check.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+changelog:
+  - type: FIX
+    description: |
```
Comment (kdorosh) [inline changelog/v1.3.13/improve-glooctl-check.yaml]: nit, prefer `>` unless you explicitly need `|` yaml formatting (which you aren't using here)

File: pkg/cliutil/uri.go [hunk 4]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
...
+
+	var errsList []string
+	for {
+		select {
+		case err := <-errs:
+			errsList = append(errsList, err.Error()+"\n")
+		case res := <-result:
+			return nil, portFwd, res
+		case <-timer:
+			return errors.Errorf("timed out trying to connect to Envoy admin port, errors: %v", errsList), portFwd, ""
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: this error message is specific in a generic function

File: pkg/cliutil/uri.go [hunk 5]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
...
+	// wait for port-forward to be ready
+	result := make(chan string)
+	errs := make(chan error)
+	go func() {
+		for {
+			res, err := http.Get("http://localhost:" + localPort + getPath)
+			if err != nil {
+				errs <- err
+				time.Sleep(time.Millisecond * 250)
+				continue
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: let's move this out and DRY up the code

File: pkg/cliutil/uri.go [hunk 6]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
...
+			}
+			res.Body.Close()
+			result <- string(b)
+			return
+		}
+	}()
+
+	timer := time.Tick(time.Second * 5)
+
+	var errsList []string
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: use `multierror` here, e.g.
```golang
	var multiErr *multierror.Error
	for _, syncer := range s {
		if err := syncer.Sync(ctx, snapshot); err != nil {
			multiErr = multierror.Append(multiErr, err)
		}
	}
	return multiErr.ErrorOrNil()
```
in the `"github.com/hashicorp/go-multierror"` package

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 7]
```diff
@@ -360,6 +368,94 @@ func checkProxies(namespaces []string) (bool, error) {
 			}
 		}
 	}
+	// check if any proxy instances are out of sync with the Gloo control plane
+	errMessage := "Problem while checking for out of sync proxies"
+
+	// port-forward gateway-proxy deployment and get prometheus metrics
+	adminPort := int(defaults.EnvoyAdminPort)
+	localPort := adminPort + 1
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go]: any reason we're making this a different port?

File: changelog/v1.3.13/improve-glooctl-check.yaml (hunk 3 above)
Comment (arianaw66) [inline changelog/v1.3.13/improve-glooctl-check.yaml]:
```suggestion
    description: >
```

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 8]
```diff
@@ -360,6 +368,94 @@ func checkProxies(namespaces []string) (bool, error) {
...
+	}
+
+	if strings.TrimSpace(stats) == "" {
+		fmt.Println(errMessage+": could not find any metrics at", promStatsPath, "endpoint of the gateway-proxy deployment")
+		return false, nil
+	}
+
+	if !strings.Contains(stats, "envoy_control_plane_connected_state{} 1") {
+		fmt.Println("Your gateway-proxy is out of sync with the Gloo control plane and is not receiving valid gloo config. " +
+			"You may want to try looking at your gloo or gateway-proxy logs or using the `glooctl debug log` command.")
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go]: could also include reference to `glooctl proxy logs` command

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 7 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: Ashley suggested it in case they/we are already port-forwarding this pod with the default port

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 8 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: This is the same as checking the gateway-proxy logs. Seemed better to keep the output short? But I'm happy to add that if you think we should in addition or instead of saying to check the gateway-proxy logs.

File: pkg/cliutil/uri.go (hunk 5 above)
Comment (arianaw66) [reply pkg/cliutil/uri.go]: What does this mean?

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 7 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/check/root.go]: We should just pick an open port then and use that, rather than hope it's open

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 8 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/check/root.go]: and `glooctl debug log` is a shortcut to the gloo logs. I think we should be consistent and return shortcuts to both logs

File: pkg/cliutil/uri.go (hunk 5 above)
Comment (kdorosh) [reply pkg/cliutil/uri.go]: ideally we rewrite the whole thing here with exponential backoff and time-limited retries, as well as a cancellable context so this goroutine can't error forever, e.g. https://github.com/solo-io/gloo/blob/master/projects/discovery/pkg/fds/updater.go#L206

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 8 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: Oh, thought I checked that. They have different output for me, but it seems they're just out of sync time-wise. Interesting. Agreed that shortcuts to both is better.

File: pkg/cliutil/uri.go [hunk 9]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
...
+				time.Sleep(time.Millisecond * 250)
+				continue
+			}
+			res.Body.Close()
+			result <- string(b)
+			return
+		}
+	}()
+
+	timer := time.Tick(time.Second * 5)
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: this timer should be built into a cancellable context that gets passed into the above go routine

File: pkg/cliutil/uri.go [hunk 10]
```diff
@@ -153,3 +154,87 @@ func minikubeIp(clusterName string) (string, error) {
...
+
+	var errsList []string
+	for {
+		select {
+		case err := <-errs:
+			errsList = append(errsList, err.Error()+"\n")
+		case res := <-result:
+			return nil, portFwd, res
+		case <-timer:
```
Comment (kdorosh) [inline pkg/cliutil/uri.go]: once we have a cancellable context, we can just change this to `ctx.Done()`

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 11]
```diff
@@ -390,6 +486,37 @@ func renderNamespaceName(namespace, name string) string {
...
 
+// parseMetrics parses prometheus metrics and returns a map from the metric name and labels to its value.
+// It expects to only look for int values!
+func parseMetrics(stats string, desiredMetricSegments []string, promStatsPath string) map[string]int {
+	statsMap := make(map[string]int)
+	statsLines := strings.Split(stats, "\n")
+	for _, line := range statsLines {
+		trimLine := strings.TrimSpace(line)
+		if strings.HasPrefix(trimLine, "#") || trimLine == "" {
+			continue // Ignore comments, help text, type info, empty lines (https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#comments-help-text-and-type-information)
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go]: any reason we're getting the prometheus-style stats and ignoring comments rather than just getting the envoy stats?

Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/check/root.go]: `/stats` vs `/stats/prometheus`

File: pkg/cliutil/uri.go [hunk 12]
```diff
@@ -4,19 +4,20 @@ import (
...
+	"io/ioutil"
 	"net/http"
 	"os"
 	"os/exec"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/solo-io/go-utils/kubeutils"
+	"github.com/solo-io/solo-kit/pkg/errors"
```
Comment (ashleywang1) [inline pkg/cliutil/uri.go]:
```suggestion
	errors "github.com/rotisserie/eris"
```

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 11 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: Because I didn't realize they were in both... :)

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 8 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/cmd/check/root.go]: It should be `glooctl debug log --errors-only`

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 11 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/root.go]: Kevin and I talked offline and decided to stick with `/stats/prometheus` because it has better documentation for the text format of its stats (https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#text-format-details), and we want to be confident that the string parsing will be safe.

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 13]
```diff
@@ -360,6 +368,97 @@ func checkProxies(namespaces []string) (bool, error) {
...
+	}
+
+	if !strings.Contains(stats, "envoy_control_plane_connected_state{} 1") {
+		fmt.Println("Your gateway-proxy is out of sync with the Gloo control plane and is not receiving valid gloo config. " +
+			"You may want to try looking at your gloo or gateway-proxy logs or using the `glooctl debug log` command.")
+		return false, nil
+	}
+
+	// wait for metrics to update
+	time.Sleep(time.Millisecond * 250)
```
Comment (ashleywang1) [inline projects/gloo/cli/pkg/cmd/check/root.go]: Put 250 * time.Millisecond into a constant called envoyRetryInterval

File: projects/gloo/cli/pkg/cmd/check/prom_stats.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+	"strconv"
+	"strings"
+)
+
+const promStatsPath = "/stats/prometheus"
+
+func checkProxyConnectedState(stats string, genericErrMessage string, connectedStateErrMessage string) bool {
+
+	if strings.TrimSpace(stats) == "" {
+		fmt.Println(genericErrMessage+": could not find any metrics at", promStatsPath, "endpoint of the gateway-proxy deployment")
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/prom_stats.go]: nit, formatting here would be off
```suggestion
		fmt.Println(genericErrMessage+": could not find any metrics at ", promStatsPath, " endpoint of the gateway-proxy deployment")
```

also would prefer `Sprintf` here too

Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/prom_stats.go]: is this going to be formatted right? do we need spaces before and after `promStatsPath`?

File: projects/gloo/cli/pkg/cmd/check/prom_stats.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+		desiredMetric := false
+		for _, s := range desiredMetricSegments {
+			if strings.Contains(trimLine, s) {
+				desiredMetric = true
+			}
+		}
+		if desiredMetric {
+			pieces := strings.Fields(trimLine) // split by white spaces
+			metric := strings.Join(pieces[0:len(pieces)-1], "")
+			metricVal, err := strconv.Atoi(pieces[len(pieces)-1])
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/prom_stats.go]: this is very complicated to read, we should pull this out into named functions and/or properly comment this parsing

File: projects/gloo/cli/pkg/cmd/check/prom_stats.go (hunk 14 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/cmd/check/prom_stats.go]: Nope, `Println` automatically adds spaces between arguments you pass it

File: test/kube2e/gateway/gateway_suite_test.go [hunk 16]
```diff
@@ -88,6 +89,9 @@ func StartTestHelper() {
 			Metadata: core.Metadata{
 				Namespace: testHelper.InstallNamespace,
 			},
+			Top: options.Top{
+				Ctx: context.Background(),
+			},
```
Comment (kdorosh) [inline test/kube2e/gateway/gateway_suite_test.go]: why this change?

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 17]
```diff
@@ -360,8 +362,9 @@ func checkProxies(namespaces []string) (bool, error) {
 			}
 		}
 	}
-	fmt.Printf("OK\n")
-	return true, nil
+
+	return checkProxyPromStats(ctx, glooNamespace)
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go]: let's make sure this command works for all possible envoy deployments, not just the gateway-proxy.

i.e, knative internal proxy, knative external proxy, ingress proxy deployment.

let's also ensure the error messages are generic (not specific to gateway-proxy)

File: test/kube2e/gateway/gateway_suite_test.go (hunk 16 above)
Comment (arianaw66) [reply test/kube2e/gateway/gateway_suite_test.go]: The gateway suite was failing without it, I think because we were trying to create a child context (with WithTimeout) on a nonexistent context.
